internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,207 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+	"gopkg.in/tomb.v2"
+)
+
+const parserSize = 4 * 1024
+
+// logGatherer is responsible for collecting service logs from a bunch of
+// services, and sending them to its logClient.
+// One logGatherer will run per log target. Its loop() method should be run
+// in its own goroutine. It can be stopped in a separate goroutine by calling
+// the stop() method.
+// The logGatherer will "flush" the client:
+//   - on a regular cadence (e.g. every 1 second)
+//   - when it is told to shut down.
+//
+// Its client may also flush itself when its internal buffer reaches a certain
+// size.
+type logGatherer struct {
+	targetName string
+	tickPeriod time.Duration
+	cancel     chan struct{}
+	client     logClient
+
+	entryCh     chan servicelog.Entry
+	pullers     map[string]*logPuller
+	pullersLock sync.Mutex
+}
+
+func newLogGatherer(target *plan.LogTarget) (*logGatherer, error) {
+	client, err := newLogClient(target)
+	if err != nil {
+		return nil, fmt.Errorf("could not create log client: %w", err)
+	}
+
+	return &logGatherer{
+		targetName: target.Name,
+		tickPeriod: 1 * time.Second,
+		cancel:     make(chan struct{}),
+		client:     client,
+		entryCh:    make(chan servicelog.Entry),
+		pullers:    map[string]*logPuller{},
+	}, nil
+}
+
+func (g *logGatherer) planChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+
+	newPullers := map[string]*logPuller{}
+
+	for _, service := range pl.Services {
+		target := pl.LogTargets[g.targetName]
+		if !service.LogsTo(target) {
+			continue
+		}
+
+		if puller, ok := g.pullers[service.Name]; ok {
+			// Move over existing puller
+			newPullers[service.Name] = puller
+			delete(g.pullers, service.Name)
+			continue
+		}
+
+		buffer, bufferExists := buffers[service.Name]
+		if !bufferExists {
+			// We don't yet have a reference to the service's ring buffer
+			// Need to wait until serviceStarted
+			continue
+		}
+
+		// Create new puller
+		newPullers[service.Name] = newLogPuller(g.entryCh, buffer)
+	}
+
+	// Old pullers for now-removed services need to be shut down.
+	for svc, puller := range g.pullers {
+		err := puller.tomb.Killf("gatherer closed")
+		if err != nil {
+			logger.Noticef("Error shutting down puller for target %q, service %q: %v",
+				g.targetName, svc, err)
+		}
+	}
+	g.pullers = newPullers
+}
+
+func (g *logGatherer) serviceStarted(service *plan.Service, buffer *servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+	g.pullers[service.Name] = newLogPuller(g.entryCh, buffer)
```

Comment:
need to check for old puller from a previous service right?

Reply:
yeah so here we need to either wait for the previous puller to die (after it reaches the end of the old ring buffer), or have the gatherer track a list of dying pullers, so it can make sure to stop them.

I think we need to seperate the lifetime of a puller from g.pullers. All pullers spawned from a gatherer should be killed when the gatherer stops, but only after they have had a chance to flush.

Reply:
`serviceStarted` will only be called for *new* services, so there is no need to check for old pullers.

Puller lifecycle is tied to the gatherer lifecycle. You're probably right that we should shut down old pullers in *planChanged* (rather than *serviceStarted*).

Reply:
A service restart actually reuses the same ringbuffer. I've updated the code to (explicitly) handle service restarts.